core/src/conn.rs

```diff
@@ -68,31 +192,59 @@ impl MongoConnection {
         Ok(connection)
     }
 
+    #[cfg(feature = "garbage_collect")]
     pub fn shutdown(self) -> Result<()> {
-        self.runtime
-            .block_on(async { self.client.shutdown().await });
+        // we need to lock the CLIENT_MAP to potentially remove the client from the map.
+        // This prevents races on the strong_count or with gc().
+        let mut client_map = self.runtime.block_on(async { CLIENT_MAP.lock().await });
+        // If this is the last reference to the Runtime, we need to shutdown the Client.
+        if Arc::strong_count(&self.runtime) == 1 {
+            self.runtime
+                .block_on(async { self.client.shutdown().await });
+        }
+        // We need to drop the Runtime before we gc, otherwise this will not be collected.
         drop(self.runtime);
+        // garbage collect any Clients and Runtimes that are no longer in use, which is denoted by
+        // the strong_count for the Runtime being 0. It's possible there could be
+        // other clients that are no longer in use because shutdown was not properly called before,
+        // so we gc them all.
+        client_map.gc();
+        Ok(())
+    }
+
+    #[cfg(not(feature = "garbage_collect"))]
+    pub fn shutdown(self) -> Result<()> {
+        // we need to lock the CLIENT_MAP to potentially remove the client from the map.
+        // This prevents races on the strong_count or with gc().
+        let client_map = self.runtime.block_on(async { CLIENT_MAP.lock().await });
+        if !client_map.0.is_empty() {
+            // we only want to shutdown if the CLIENT_MAP has not been used.
```

Comment:
As soon as a connection is made with OIDC, there will be an entry in the map, no?
I think I'm missing something because the way I read this, if a user makes an OIDC connection, then shutdown will never be called and I thought never calling shutdown was what was causing SSIS to hang.

Reply:
Yes, this will make it not shutdown if they are using OIDC. Now I'm thinking we may want to pool the clients even if they are not using OIDC given the thing with tableau server we saw and JDBC. And no, not calling shutdown was not what was causing SSIS to hang, the runtime will still be shutdown when the driver is unloaded, and since it's the current thread runtime, there will be no problem with hanging.